core/src/main/java/org/apache/calcite/sql/fun/SqlBasicAggFunction.java

```diff
@@ -181,7 +184,21 @@ public SqlBasicAggFunction withAllowsSeparator(boolean allowsSeparator) {
         getReturnTypeInference(), getOperandTypeInference(),
         getOperandTypeChecker(), getFunctionType(), requiresOrder(),
         requiresOver(),  requiresGroupOrder(), distinctOptionality, syntax,
-        allowsNullTreatment, allowsSeparator);
+        allowsNullTreatment, allowsSeparator, isPercentile);
+  }
+
+  /** Returns whether this aggregate function is a PERCENTILE function. */
+  public boolean isPercentile() {
```

Comment:
Move this method to the base class. (It can have a default 'false' in the base class.)

Then we can skip the `instanceof` and cast before calling this function.

Reply:
This is not in the latest patch anymore. I have moved all the code into a separate visitor which does not use any downcasting/instance checking.